scarb/src/compiler/compilers/starknet_contract/artifacts_writer.rs

```diff
@@ -0,0 +1,150 @@
+use cairo_lang_defs::ids::NamedLanguageElementId;
+use cairo_lang_starknet::contract::ContractDeclaration;
+use cairo_lang_starknet_classes::casm_contract_class::CasmContractClass;
+use cairo_lang_starknet_classes::contract_class::ContractClass;
+use cairo_lang_utils::UpcastMut;
+use itertools::{izip, Itertools};
+use scarb_stable_hash::short_hash;
+use serde::Serialize;
+use smol_str::SmolStr;
+
+#[derive(Debug, Serialize)]
+struct StarknetArtifacts {
+    version: usize,
+    contracts: Vec<ContractArtifacts>,
+}
+
+impl Default for StarknetArtifacts {
+    fn default() -> Self {
+        Self {
+            version: 1,
+            contracts: Vec::new(),
+        }
+    }
+}
+
+impl StarknetArtifacts {
+    fn finish(&mut self) {
+        assert!(
+            self.contracts.iter().map(|it| &it.id).all_unique(),
+            "Artifacts IDs must be unique."
+        );
+
+        self.contracts.sort_unstable_by_key(|it| it.id.clone());
+    }
+}
+
+#[derive(Debug, Serialize)]
+struct ContractArtifacts {
+    id: String,
+    package_name: PackageName,
+    contract_name: String,
+    module_path: String,
+    artifacts: ContractArtifact,
+}
+
+impl ContractArtifacts {
+    fn new(
+        package_name: &PackageName,
+        contract_name: &str,
+        contract_path: &str,
+        module_path: &str,
+    ) -> Self {
+        Self {
+            id: short_hash((&package_name, &contract_path)),
+            package_name: package_name.clone(),
+            contract_name: contract_name.to_owned(),
+            module_path: module_path.to_owned(),
+            artifacts: ContractArtifact::default(),
+        }
+    }
+}
+
+#[derive(Debug, Default, Serialize)]
+struct ContractArtifact {
+    sierra: Option<String>,
+    casm: Option<String>,
+}
+
+pub struct ArtifactsWriter {
+    sierra: bool,
+    casm: bool,
+    target_dir: Filesystem,
+    target_name: SmolStr,
+}
+
+impl ArtifactsWriter {
+    pub fn new(target_name: SmolStr, target_dir: Filesystem, props: Props) -> Self {
+        Self {
+            sierra: props.sierra,
+            casm: props.casm,
+            target_dir,
+            target_name,
+        }
+    }
+
+    pub fn write(
+        self,
+        contract_paths: Vec<String>,
+        contracts: &Vec<ContractDeclaration>,
+        classes: &[ContractClass],
+        casm_classes: &[Option<CasmContractClass>],
+        db: &mut RootDatabase,
+        ws: &Workspace<'_>,
+    ) -> anyhow::Result<()> {
+        let mut artifacts = StarknetArtifacts::default();
+        let mut file_stem_calculator = ContractFileStemCalculator::new(contract_paths);
+
+        for (decl, class, casm_class) in izip!(contracts, classes, casm_classes) {
+            let contract_name = decl.submodule_id.name(db.upcast_mut());
+            let contract_path = decl.module_id().full_path(db.upcast_mut());
+
+            let contract_selector = ContractSelector(contract_path);
+            let package_name = contract_selector.package();
```

Comment:
With the current implementation it seems that compiling a given conract with `test` and `starknet-contract` targets will produce artifacts with the same `contract_name` and `package_name` (and `id` as well?). Would it make sense to distinguish between these artifacts at this level, perhaps by appending `unittest` or `integrationtest` to the contract name?

Reply:
Actually, the contract names are not prepended by package name, but rather by the target name ( https://github.com/software-mansion/scarb/blob/1aa3ca598e12bd692c874f11cb7e87b54cf125e3/scarb/src/compiler/compilers/starknet_contract/artifacts_writer.rs#L112), so this should not overwrite artifacts from other targets. 

![image](https://github.com/user-attachments/assets/9eaa0c42-086a-44a3-9d71-9efe03e12587)